src/test/connect-test.c

```diff
@@ -0,0 +1,444 @@
+        return -1;
+}
+
+static int server_handler_4(sd_event_source *event_source, int fd, uint32_t revents, void *data) {
+        return server_handler(event_source, fd, revents, data);
+}
+
+static int server_handler_6(sd_event_source *event_source, int fd, uint32_t revents, void *data) {
+        return server_handler(event_source, fd, revents, data);
+}
+
+static int conn_handler(sd_event_source *event_source, int fd, uint32_t revents, void *data) {
+        int error, r;
+        struct context *ctx = data;
+        unsigned int length = sizeof(error);
+        struct sockaddr addr = { 0 };
+        socklen_t addr_length = sizeof(addr);
+        enum connection_status *status;
+
+        r = getsockname(fd, &addr, &addr_length);
+        if (r < 0) {
+                sd_event_exit(sd_event_source_get_event(event_source), r);
+                return -1;
+        }
+
+        if (addr.sa_family == AF_INET)
+                status = &(ctx->ipv4);
+        else if (addr.sa_family == AF_INET6)
+                status = &(ctx->ipv6);
+        else {
+                sd_event_exit(sd_event_source_get_event(event_source), -1);
+                return -1;
+        }
+
+        r = getsockopt(fd, SOL_SOCKET, SO_ERROR, &error, &length);
+        if (r < 0) {
+                sd_event_exit(sd_event_source_get_event(event_source), r);
+                return -1;
+        }
+
+        if (error == 0)
+                /* connection succeeded */
+                *status = CONNECTION_SUCCEEDED;
+        else if (error == ECONNREFUSED)
+                /* conncetion succeeded but no-one was listening */
+                *status = CONNECTION_REFUSED;
+        else if (error == ETIMEDOUT)
+                *status = CONNECTION_TIMEOUT;
+        else {
+                sd_event_exit(sd_event_source_get_event(event_source), -1);
+                return -1;
+        }
+
+        if (ctx->ipv4 != CONNECTION_UNDEFINED && ctx->ipv6 != CONNECTION_UNDEFINED)
+                sd_event_exit(sd_event_source_get_event(event_source), 0);
+
+        return -1;
+}
+
+static int conn_handler_4(sd_event_source *event_source, int fd, uint32_t revents, void *data) {
+        return conn_handler(event_source, fd, revents, data);
+}
+
+static int conn_handler_6(sd_event_source *event_source, int fd, uint32_t revents, void *data) {
+        return conn_handler(event_source, fd, revents, data);
+}
+
+static void usage(const char *program_name) {
+        fprintf(stderr,
+                "Usage: %s <port> <timeout_ms> <"INPUT_CONNECTION_SUCCESS"|"INPUT_CONNECTION_REJECTED"|"INPUT_CONNECTION_TIMEOUT">\n",
+                program_name);
+}
+
+int main(int argc, char *argv[]) {
+        sd_event *e = NULL;
+        sd_event_source *event_source = NULL;
+        int r = -1, fd_4 = 0, fd_6 = 0, listen_fd_4 = 0, listen_fd_6 = 0, reuseaddr = 1;
+        struct sockaddr_in addr_4 = { 0 }, serv_addr_4 = { 0 };
+        struct sockaddr_in6 addr_6 = { 0 }, serv_addr_6 = { 0 };
```

Comment:
even though not strictly ANSI C we generally prefer initializing structs to zero simply by using `{}`.